firebase_admin/db.py

```diff
@@ -37,629 +36,644 @@
+		"""
+		if value is None:
+			raise ValueError('Value must not be None.')
+		output = self._client.request('post', self._add_suffix(), json=value)
+		push_id = output.get('name')
+		return self.child(push_id)
+
+	def update(self, value):
+		"""Updates the specified child keys of this Reference to the provided values.
+		Args:
+		  value: A dictionary containing the child keys to update, and their new values.
+		Raises:
+		  ValueError: If value is empty or not a dictionary.
+		  ApiCallError: If an error occurs while communicating with the remote database server.
+		"""
+		if not value or not isinstance(value, dict):
+			raise ValueError('Value argument must be a non-empty dictionary.')
+		if None in value.keys() or None in value.values():
+			raise ValueError('Dictionary must not contain None keys or values.')
+		self._client.request_oneway('patch', self._add_suffix(), json=value, params='print=silent')
+
+	def update_with_etag(self, value, etag):
+		"""Updates the specified child keys of this Reference to the provided values and uses ETag to make sure data is up to date.
+		Args:
+			value: A dictionary containing the child keys to update, and their new values.
+			etag: ETag value for the Reference.
+		Returns:
+			ValueError: If value is empty or not a dictionary, or if etag is not a string. 
+		"""
+		if not value or not isinstance(value, dict):
+			raise ValueError('Value argument must be a non-empty dictionary.')
+		if None in value.keys() or None in value.values():
+			raise ValueError('Dictionary must not contain None keys or values.')
+		if not isinstance(etag, str):
+			raise ValueError('ETag must be a string.')
+
+		try:
+			self._client.request_oneway('put', self._add_suffix(), json=value, headers={'if-match': etag})
+		except ApiCallError as error:
+			detail = error.detail
+			snapshot = detail.response.json()
+			etag = detail.response.headers['ETag']
+			return etag, snapshot
+
+	def delete(self):
+		"""Deleted this node from the database.
+		Raises:
+		  ApiCallError: If an error occurs while communicating with the remote database server.
+		"""
+		self._client.request_oneway('delete', self._add_suffix())
+
+	def transaction(self, transaction_update, on_complete=None):
+		"""Write to database using a transaction.
+		Args:
+			transaction_update: function that takes in current database data as a parameter.
+			on_complete: function that takes takes in the following parameters:
+				error: Error message, possibly null
+				committed: Whether the transaction_update function committed data to the database
+				data: The data currently in the database
+
+		"""
+		if not callable(transaction_update):
+			raise ValueError('transaction_update must be a function.')
+		if on_complete is not None and not callable(on_complete):
+			raise ValueError('on_complete must be a function.')
+
+		error = None
+		committed = False
+		try:
+			tries = 0
+			etag, data = self.get_with_etag()
+			val = transaction_update(data)
```

Comment:
Hey Alexander, this looks awesome! The one thing I'd say is that you could change transaction_update to return a tuple of (val, retry) where retry is a boolean. Then, you can change you while loop to be:
  while retry == True:
This way the user passes in a transaction_update function that controls retry logic itself. We could also still include _FIREBASE_MAX_RETRIES so that users don't accidentally infinite loop, though I believe our SDKs that already have transactions do not do have this-so it may be best to be consistent.